README.md
# Introduction

This document tries to answer some questions you might have about Jasmin. In particular, several people have asked me what Jasmin is, why they might use Jasmin, and why I wrote it in the first place. I've tried to give some answers to these questions below.

# What is Jasmin?

Jasmin is a Java Assembler Interface. It takes ASCII descriptions for Java classes, written in a simple assembler-like syntax using the Java Virtual Machine instructions set. It converts them into binary Java class files suitable for loading into a Java interpreter.

To give you a flavor, here is the Jasmin assembly code for HelloWorld:

```jasmin
.class public HelloWorld
.super java/lang/Object

;
; standard initializer (calls java.lang.Object's initializer)
;
.method public <init>()V
   aload_0
   invokenonvirtual java/lang/Object/<init>()V
   return
.end method

;
; main() - prints out Hello World
;
.method public static main([Ljava/lang/String;)V
   .limit stack 2   ; up to two items can be pushed
       ; push System.out onto the stack
   getstatic java/lang/System/out Ljava/io/PrintStream;

   ; push a string onto the stack
   ldc "Hello World!"

   ; call the PrintStream.println() method.
   invokevirtual java/io/PrintStream/println(Ljava/lang/String;)V

   ; done
   return
.end method
```

Jasmin was originally created as a companion to the book "Java Virtual Machine", written by Jon Meyer and Troy Downing and published by O'Reilly Associates. The book is now out of print. Jasmin survives as a SourceForge Open Source project.

# Motivation for Jasmin

Jasmin was written because, at the time that we wrote the Java Virtual Machine book for O'Reilly, Sun had not published an assembler format for the Java virtual machine.

Generating a binary Java .class file is pretty fiddly. Its like creating an a.out (or .exe) file by hand. Even using a Java package like JAS (a Java API for creating class files, used internally by Jasmin and written by KB Sriram), you need to know a lot about the philosophy of the Java Virtual Machine before you can write something at the Virtual Machine level and generate a Java class.

We wanted something that made it very easy for a student or programmer to explore the Java Virtual Machine, or write a new language which targets the VM, without getting into the details of constant pool indices, attribute tables, and so on.

Creating a Java assembler seemed like a good solution.

Unfortunately, Sun has not seen the need to define a standard Java assembler format, and has not released tools perform Java assembly.

Sun does provide a javap program which can print the assembly code in a class file. However, the javap output is inappropriate for use as an assembler format. It is designed to be read by a person, not to be parsed by an assembler, so it has a number of omissions and drawbacks.

Internally, Sun has a Java assembler tool. In hindsight, the syntax used by their internal tool is nicer than the Jasmin syntax. However, to my knowledge, their tool is still not widely available, nor is it a formally supported part of the Sun JDK.

# Update on Jasmin Today (2004)

Since Jasmin was written, it has become the de-facto standard assembly format for Java. It is used in dozens of compiler classes throughout the world, and has been ported and cloned multiple times. For better or worse, Jasmin remains the original Java assembler.

[As an interesting comparison, Microsoft .NET shipped out-of-box with an assembler, a disassembler, a standard IL assembly format, and built-in libraries for code-genning (generating classes on the fly). It would be great if Sun was as comprehensive in their support of the JVM].

# What can I do with Jasmin?

To give you some ideas, below are some theoretical Jasmin users/uses.

### Teachers

If you are teaching a compilers course, you could have students write a compiler which generates Jasmin assembly files, and then assembles those files into Java class files. Then you can integrate the advantages of the Virtual Machine (portability, the verifier, an object model...) into your courseware.

### Hobbyists

Jasmin lets you poke around in Java at the VM level, so that you can gain a real understanding of how Java works and what the Virtual Machine is like.

### System Implementors

If you are implementing a Java runtime system, Jasmin is an essential tool for generating test classes.

### Advanced Programmers

You could use Jasmin to write a critical class or method by hand (e.g. if you think that Java isn't doing things as well as it could).

Alternatively, you could create a syntax extension to the Java language which uses Jasmin (or JAS).

### Language Implementors

If you want to create an implementation of your favorite programming language which targets the Virtual Machine, Jasmin may be a simpler approach than writing a Java class file generator. This is especially true if your compiler is implemented in something other than Java, since you can create Java class files easily without having to get involved in the details of the binary file format.

### Security Wizards

Sun's claim that the Java class verifier protects you from hostile programs is a pretty strong one. Jasmin lets you create 'hostile' class files and see if a Java implementation is really as secure as it should be.

*Copyright (c) Jonathan Meyer, July 1996*

http://jasmin.sf.net/about.html


changes.txt
Jasmin Revision History                                     Jonathan Meyer


02 Dec 2005 - Release 2.0 by Daniel Reynaud

   * defined the JasminXT language, see the documentation for more details



15 Oct 2004 - Release 1.1

   * Switched to Ant Build System

   * Moved java_cup and jas sources into src directory

   * Updated documentation to use style sheets

   * Changed docs to reflect using jar files rather than class files

   * Uploaded to SourceForge


-----------------------------------------------------------------------------

11 Apr 97 - Release 1.06.

11 Apr 97

   * Fixed bug which prevented the source name from being written out
     in the class file

   * Improved README file

2 Mar 97 - Release 1.05.

1 Mar 97

   * Moved scripts into a bin directory. 

   * Added support for Visual J++.

     Added vjasmin.bat, for running Jasmin using Visual J++.

     Converted JAS/Jasmin to use its own internal RuntimeConstants, so that
     there is no longer any dependency on Sun's version (needed by J++).

   * Tidied API:

     Renamed "Jasmin" class "ClassFile" (sorry to those of you using the API
     from 1.04). The ClassFile class is documented in the doc/api directory.

   * Mods for Java 1.1:

     Classes now set the ACC_SUPER bit in their access flags.
     i2b/i2s/i2c are now synonyms for int2byte, int2short, int2char.
     invokespecial is now a synonym for invokenonvirtual.

   * Mods to pick up documentation in book:
    
     "wide" is now a recognized instruction in Jasmin files - although the assembler
     just ignores it!
 
     Added the optional <high> parameter to tableswitch. 

   * Fixed bug in .catch all
    
10 Feb 96 - Release 1.04.

8 Feb 97
   * Updated to use latest version of JAS. This fixes some bugs in the
     earlier release (including handling of _w instructions)

   * Split several of the internal classes into smaller pieces.

   * Restructured internal sources so that Jasmin, Scanner and parser
     no longer rely on static data structures. Now there is a public API
     to Jasmin, for people that want to assemble classes using their own
     data input/output streams.

30 Oct 96

   * Added support for more \ escapes in quoted strings. In
     particular, you can now use \nnn to specify a character using
     octal.

2 Oct 96 - Release 1.03.

1 Oct 96

   * Added better support for interfaces: added the .interface
     directive (an alternative to the .class directive), and also a
     .implements directive. Updates guide.html to mention these new
     features.

24 Sept 96

   * Fixed several problems with guide.html - thanks to feedback from
     Shawn Silverman (umsilve1@cc.umanitoba.ca).

23 Aug 96

   * Tidied up documentation and implementation for wide instructions.

     Now ldc and ldc_w are used for single-word items, whereas
     ldc2_w is used for two word items (previously, I had ldc_w as
     a synonym for ldc2_w - oops).

25 July 96

   * Added documentation for .var directive.

   * Fixed line numbering produced by -g flag (I hope).

   * Improved error reporting slightly.

24 July 96

   * Added fix to scanner to handle Ctrl-M characters,
     for DOS/NT Systems. (Thanks sbk!)

18 July 96 - Release 1.0.


src/jasmin/sym.java

//----------------------------------------------------
// The following code was generated by Java(tm) CUP v0.9d
// Fri Dec 02 14:52:27 CET 2005
//----------------------------------------------------

package jasmin;

/** JavaCup generated class containing symbol constants. */
public class sym {
  /* terminals */
  static final int ANNOTATION = 41;
  static final int SEP = 49;
  static final int TABLESWITCH = 46;
  static final int DTHROWS = 12;
  static final int DCATCH = 2;
  static final int USING = 22;
  static final int PRIVATE = 34;
  static final int DSOURCE = 11;
  static final int INTERFACE = 32;
  static final int DBYTECODE = 16;
  static final int STATIC = 37;
  static final int LOOKUPSWITCH = 45;
  static final int IS = 23;
  static final int DINTERFACE = 15;
  static final int Str = 51;
  static final int OFFSET = 28;
  static final int DFIELD = 5;
  static final int Int = 54;
  static final int DCLASS = 3;
  static final int VARARGS = 44;
  static final int ABSTRACT = 30;
  static final int EOF = 0;
  static final int METHOD = 25;
  static final int TRANSIENT = 39;
  static final int error = 1;
  static final int Insn = 53;
  static final int STACK = 27;
  static final int Word = 52;
  static final int DSUPER = 10;
  static final int DLIMIT = 6;
  static final int EQ = 48;
  static final int Signed = 56;
  static final int Num = 55;
  static final int DIMPLEMENTS = 14;
  static final int FROM = 24;
  static final int DDEBUG = 17;
  static final int DSTACK = 20;
  static final int DMETHOD = 8;
  static final int COLON = 50;
  static final int NATIVE = 33;
  static final int PROTECTED = 35;
  static final int ENUM = 42;
  static final int TO = 21;
  static final int DEND = 4;
  static final int FINAL = 31;
  static final int DEFAULT = 47;
  static final int BRIDGE = 43;
  static final int PUBLIC = 36;
  static final int DENCLOSING = 18;
  static final int VOLATILE = 40;
  static final int SIGNATURE = 26;
  static final int DVAR = 13;
  static final int SYNCHRONIZED = 38;
  static final int DSET = 9;
  static final int LOCALS = 29;
  static final int DLINE = 7;
  static final int DSIGNATURE = 19;
};

